File: vector/src/main/java/im/vector/activity/WidgetActivity.java
/*
 * Copyright 2017 Vector Creations Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package im.vector.activity;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.PermissionRequest;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.TextView;

import org.matrix.androidsdk.MXSession;
import org.matrix.androidsdk.data.Room;
import org.matrix.androidsdk.rest.callback.ApiCallback;
import org.matrix.androidsdk.rest.model.MatrixError;
import org.matrix.androidsdk.util.Log;

import butterknife.BindView;
import butterknife.ButterKnife;
import im.vector.Matrix;
import im.vector.R;
import im.vector.widgets.Widget;
import im.vector.widgets.WidgetsManager;

/*
 * This class displays a widget
 */
public class WidgetActivity extends RiotAppCompatActivity {
    private static final String LOG_TAG = WidgetActivity.class.getSimpleName();

    /**
     * The linked widget
     */
    public static final String EXTRA_WIDGET_ID = "EXTRA_WIDGET_ID";

    // the linked widget
    private Widget mWidget = null;

    // the session
    private MXSession mSession;

    // the room
    private Room mRoom;

    @BindView(R.id.back_icon_container)
    View mBackToAppIcon;

    @BindView(R.id.close_widget_icon_container)
    View mCloseWidgetIcon;

    @BindView(R.id.widget_web_view)
    WebView mWidgetWebView;

    @BindView(R.id.widget_type_text_view)
    TextView mWidgetTypeTextView;

    @BindView(R.id.widget_progress_layout)
    View waitingView;

    /**
     * Widget events listener
     */
    private final WidgetsManager.onWidgetUpdateListener mWidgetListener = new WidgetsManager.onWidgetUpdateListener() {
        @Override
        public void onWidgetUpdate(Widget widget) {
            if (TextUtils.equals(widget.getWidgetId(), mWidget.getWidgetId())) {
                if (!widget.isActive()) {
                    WidgetActivity.this.finish();
                }
            }
        }
    };

    @Override
    @SuppressLint("NewApi")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_widget);
        ButterKnife.bind(this);

        mWidget = (Widget) getIntent().getSerializableExtra(EXTRA_WIDGET_ID);

        if ((null == mWidget) || (null == mWidget.getUrl())) {
            Log.e(LOG_TAG, "## onCreate() : invalid widget");
            finish();
            return;
        }

        mSession = Matrix.getMXSession(this, mWidget.getSessionId());

        if (null == mSession) {
            Log.e(LOG_TAG, "## onCreate() : invalid session");
            finish();
            return;
        }

        mRoom = mSession.getDataHandler().getRoom(mWidget.getRoomId());

        if (null == mRoom) {
            Log.e(LOG_TAG, "## onCreate() : invalid room");
            finish();
            return;
        }

        mWidgetTypeTextView.setText(mWidget.getHumanName());
        refreshStatusBar();
        loadURL();
    }

    /**
     * Refresh the status bar
     */
    private void refreshStatusBar() {
        boolean canCloseWidget = (null == WidgetsManager.getSharedInstance().checkWidgetPermission(mSession, mRoom));

        // close widget button
        mCloseWidgetIcon.setVisibility(canCloseWidget ? View.VISIBLE : View.GONE);
        mCloseWidgetIcon.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new AlertDialog.Builder(WidgetActivity.this)
                        .setMessage(R.string.widget_delete_message_confirmation)
                        .setPositiveButton(R.string.remove, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                                showWaitingView();
                                WidgetsManager.getSharedInstance().closeWidget(mSession, mRoom, mWidget.getWidgetId(), new ApiCallback<Void>() {
                                    @Override
                                    public void onSuccess(Void info) {
                                        WidgetActivity.this.finish();
                                    }

                                    private void onError(String errorMessage) {
                                        stopWaitingView();
                                        CommonActivityUtils.displayToast(WidgetActivity.this, errorMessage);
                                    }

                                    @Override
                                    public void onNetworkError(Exception e) {
                                        onError(e.getLocalizedMessage());
                                    }

                                    @Override
                                    public void onMatrixError(MatrixError e) {
                                        onError(e.getLocalizedMessage());
                                    }

                                    @Override
                                    public void onUnexpectedError(Exception e) {
                                        onError(e.getLocalizedMessage());
                                    }
                                });
                            }
                        })
                        .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                            }
                        })
                        .create()
                        .show();
            }
        });

        mBackToAppIcon.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mWidgetWebView.canGoBack()) {
                    mWidgetWebView.goBack();
                } else {
                    WidgetActivity.this.finish();
                }
            }
        });
    }

    /**
     * Load the widget call
     */
    @SuppressLint("NewApi")
    private void loadURL() {

        // xml value seems ignored
        mWidgetWebView.setBackgroundColor(0);

        // clear caches
        mWidgetWebView.clearHistory();
        mWidgetWebView.clearFormData();
        mWidgetWebView.clearCache(true);

        WebSettings settings = mWidgetWebView.getSettings();

        // does not cache the data
        settings.setCacheMode(WebSettings.LOAD_NO_CACHE);

        // Enable Javascript
        settings.setJavaScriptEnabled(true);

        // Use WideViewport and Zoom out if there is no viewport defined
        settings.setUseWideViewPort(true);
        settings.setLoadWithOverviewMode(true);

        // Enable pinch to zoom without the zoom buttons
        settings.setBuiltInZoomControls(true);

        // Allow use of Local Storage
        settings.setDomStorageEnabled(true);

        settings.setAllowFileAccessFromFileURLs(true);
        settings.setAllowUniversalAccessFromFileURLs(true);

        settings.setDisplayZoomControls(false);

        // Permission requests
        mWidgetWebView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onPermissionRequest(final PermissionRequest request) {
                new Handler(Looper.getMainLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        request.grant(request.getResources());
                    }
                });
            }
        });

        mWidgetWebView.setWebViewClient(new WebViewClient());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            android.webkit.CookieManager cookieManager = android.webkit.CookieManager.getInstance();
            cookieManager.setAcceptThirdPartyCookies(mWidgetWebView, true);
        }

        showWaitingView();
        WidgetsManager.getFormattedWidgetUrl(this, mWidget, new ApiCallback<String>() {
            @Override
            public void onSuccess(String url) {
                stopWaitingView();
                mWidgetWebView.loadUrl(url);
            }

            private void onError(String errorMessage) {
                CommonActivityUtils.displayToast(WidgetActivity.this, errorMessage);
                WidgetActivity.this.finish();
            }

            @Override
            public void onNetworkError(Exception e) {
                onError(e.getLocalizedMessage());
            }

            @Override
            public void onMatrixError(MatrixError e) {
                onError(e.getLocalizedMessage());
            }

            @Override
            public void onUnexpectedError(Exception e) {
                onError(e.getLocalizedMessage());
            }
        });
    }


    @Override
    protected void onPause() {
        super.onPause();
        mWidgetWebView.pauseTimers();
        mWidgetWebView.onPause();

        WidgetsManager.removeListener(mWidgetListener);
    }

    /**
     * Force to render the activity in fullscreen
     */
    private void displayInFullScreen() {
        getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
    }

    @Override
    protected void onResume() {
        super.onResume();
        displayInFullScreen();
        WidgetsManager.addListener(mWidgetListener);

        mWidgetWebView.resumeTimers();
        mWidgetWebView.onResume();

        refreshStatusBar();
    }

    @Override
    protected void onDestroy() {
        if (null != mWidgetWebView) {
            ((ViewGroup) (mWidgetWebView.getParent())).removeView(mWidgetWebView);
            mWidgetWebView.removeAllViews();
            mWidgetWebView.destroy();
            mWidgetWebView = null;
        }
        super.onDestroy();
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            displayInFullScreen();
        }
    }
}

File: vector/src/main/java/im/vector/activity/WidgetActivity.kt
/*
 * Copyright 2017 Vector Creations Ltd
 * Copyright 2018 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package im.vector.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import com.airbnb.mvrx.viewModel
import im.vector.R
import im.vector.fragments.roomwidgets.*
import im.vector.ui.themes.ThemeUtils
import im.vector.widgets.Widget

/*
 * This class displays a widget
 */
class WidgetActivity : VectorAppCompatActivity() {

    val viewModel: RoomWidgetViewModel by viewModel()

    /* ==========================================================================================
     * LIFE CYCLE
     * ========================================================================================== */

    override fun getLayoutRes() = R.layout.activity_widget

    override fun getMenuRes() = R.menu.menu_room_widget

    override fun getTitleRes() = R.string.room_widget_activity_title

    @SuppressLint("NewApi")
    override fun initUiAndData() {
        configureToolbar()

        supportActionBar?.setHomeAsUpIndicator(ContextCompat.getDrawable(this, R.drawable.ic_material_leave)?.let {
            ThemeUtils.tintDrawableWithColor(it, Color.WHITE)
        })

        viewModel.selectSubscribe(this, RoomWidgetViewModelState::status) { ws ->
            when (ws) {
                WidgetState.UNKNOWN            -> {
                }
                WidgetState.WIDGET_NOT_ALLOWED -> {
                    val dFrag = supportFragmentManager.findFragmentByTag(FRAGMENT_TAG_PERMISSION) as? RoomWidgetPermissionBottomSheet
                    if (dFrag != null && dFrag.dialog?.isShowing == true && !dFrag.isRemoving) {
                        //already there
                    } else {
                        RoomWidgetPermissionBottomSheet
                                .newInstance(viewModel.session!!.myUserId, viewModel.widget).apply {
                                    onFinish = { accepted ->
                                        if (!accepted) finish()
                                    }
                                }
                                .show(supportFragmentManager, FRAGMENT_TAG_PERMISSION)
                    }
                }
                WidgetState.WIDGET_ALLOWED     -> {
                    //mount the webview fragment if needed
                    if (supportFragmentManager.findFragmentByTag(FRAGMENT_TAG_WEBVIEW) == null) {
                        supportFragmentManager.beginTransaction()
                                .replace(R.id.fragment_container, RoomWidgetFragment(), FRAGMENT_TAG_WEBVIEW)
                                .commit()
                    }
                }
            }
        }

        viewModel.selectSubscribe(this, RoomWidgetViewModelState::widgetName) { name ->
            supportActionBar?.title = name
        }

        viewModel.selectSubscribe(this, RoomWidgetViewModelState::canManageWidgets) {
            invalidateOptionsMenu()
        }

        viewModel.navigateEvent.observe(this, Observer { uxStateEvent ->
            when (uxStateEvent?.getContentIfNotHandled()) {
                RoomWidgetViewModel.NAVIGATE_FINISH -> {
                    finish()
                }
            }
        })

        viewModel.toastMessageEvent.observe(this, Observer {
            it?.getContentIfNotHandled()?.let {
                Toast.makeText(this, it, Toast.LENGTH_LONG).show()
            }
        })
    }

    /* ==========================================================================================
     * companion
     * ========================================================================================== */

    companion object {

        private const val FRAGMENT_TAG_PERMISSION = "FRAGMENT_TAG_PERMISSION"
        private const val FRAGMENT_TAG_WEBVIEW = "WebView"
        /**
         * The linked widget
         */
        const val EXTRA_WIDGET_ID = "EXTRA_WIDGET_ID"

        fun getIntent(context: Context, widget: Widget): Intent {
            return Intent(context, WidgetActivity::class.java)
                    .apply {
                        putExtra(EXTRA_WIDGET_ID, widget)
                    }
        }
    }
}